### ISSUE TITLE
Rule Enhancement Proposal: autofix for prefer-template 

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6978]
`prefer-template` is warning concatenations of between strings and expressions.
If the rule supports autofix, it's very helpful to us since we need many steps to fix the code.

1. If there is \` or `${` in the string, escapes it: 
    - \` → \\\`
    - `${` → `\${`
2. Replaces the concatenation by a template curly:
    - `" + expression + "` → `${expression}`
    - `" + expression` → `${expression}"`
    - `expression + "` → `"${expression}`
    - If there are line terminators around the `+`s, should keep those to inside of template curlies.
    - If the `expression` is enclosed parentheses, should remove the parentheses.
        - `" + (a + 1) + "` → `${a + 1}`
3. Replaces the enclosing `"`/`'` by \`.

Examples:

```js
"hello, " + name + "."
// ↓
`hello, ${name}.`
```

```js
"hello, " + name
// ↓
`hello, ${name}`
```

```js
"favorites: " + favorites.map(f => {
    return f.name
}) + ";"
// ↓
`favorites: ${favorites.map(f => {
    return f.name
})};`
```

```js
"favorites: " + 
    favorites.map(f => {
        return f.name
    }) + 
    ";"
// ↓
`favorites: ${
    favorites.map(f => {
        return f.name
    })
};`
```

**Comment by platinumazure** [comment]
A fixer for this rule will have a similar problem to comma-style in that intervening comments make fixing very difficult. Personally I would suggest just not fixing those cases.

**Comment by mysticatea** [comment]
@platinumazure I think that keeping comments is not difficult (https://github.com/eslint/eslint/pull/6957#issuecomment-242562708) :smile:

**Comment by kaicataldo** [comment]
:+1:

**Comment by not-an-aardvark** [comment]
I can work on this (assuming @mysticatea isn't working on it already).

**Comment by not-an-aardvark** [comment]
What should the fixer do with comments in this case? There are no brackets where the comments can be placed.

```js
var foo = bar + 'baz' + /* comment */ 'qux';
```

**Comment by platinumazure** [comment]
@not-an-aardvark Unless someone with more experience than I suggests a better plan, I'd say let's just not auto-fix those cases. We can always add that functionality in later in the worst case.

**Comment by mysticatea** [comment]
@not-an-aardvark Thank you for implementing! `prefer-template` does not warn concatenations between string literals. it's warned by [no-useless-concat](http://eslint.org/docs/rules/no-useless-concat).

**Comment by not-an-aardvark** [comment]
I see; so should `prefer-template` just do this fix?

```js
var foo = bar + 'baz' + 'qux';

// gets fixed to:

var foo = `${bar}baz` + 'qux';
```

?

**Comment by mysticatea** [comment]
Yes.

**Comment by not-an-aardvark** [comment]
Looking at the examples, it seems inconsistent to do this:

```js
"hello, " + name + "."
// ↓
`hello, ${name}.` // the whitespace before and after the + sign is removed

"hello, " +
  name + "."
// ↓
`hello, ${
  name}.` // the newline after the + sign is not removed
```

Instead, maybe it would be better to keep all the whitespace around the plus sign:

```js
"hello, " + name + "."
// ↓
`hello, ${  name  }.`
```

Then [`template-curly-spacing`](http://eslint.org/docs/rules/template-curly-spacing) can fix the whitespace if necessary.

**Comment by mysticatea** [comment]
> Instead, maybe it would be better to keep all the whitespace around the plus sign:

I definitely agree.
I'm sorry for that inconsistency.

Thank you!

**Comment by mysticatea** [comment]
Closing in favor of #7165.
